From 49b325144942068066a9b56dae2fd29f704c2f99 Mon Sep 17 00:00:00 2001
From: Alexandre Rostovtsev <tetromino@gmail.com>
Date: Sat, 21 May 2011 04:41:18 -0400
Subject: [PATCH] Build gnomesegvhandler module for gtk-2 and gtk-3

This patch allows bug-buddy's crash handler module to be built both for
gtk-2 and 3. To test, export GTK_MODULES=gnomesegvhandler and make a
gtk (2 or 3) application crash; if bug-buddy is installed, it will run
and handle the crash.

Details:
* configure will take --enable-gtk2-segv-handler and --enable-gtk3-segv-handler
  options. By default, both options are "auto" (if the appropriate
  version of gdk is detected, the module for that version of gtk will be
  built), but can be set to "yes" or "no" to override that behavior.
* Both modules have the same name (libgnomesegvhandler, same as
  bug-buddy uses currently). They are built in separate subdirectories
  under segv_handler.
* The GDK_DISPLAY() macro in gnome-segvhanlder.c has been replaced with
  gdk_x11_get_default_xdisplay(), which AFAIK has been the recommended
  replacement for that macro since the first release of gdk-2. This is
  necessary since in gdk-3, GDK_DISPLAY is finally gone for good.
* Removed a comment about a long-abandoned idea that does not make sense
  (550135, resolved as NOTABUG in 2009).

Signed-off-by: Alexandre Rostovtsev <tetromino@gmail.com>
---
 configure.in                     |   54 +++++++++++++++++++++++++++++++++-----
 segv_handler/Makefile.am         |   28 +++++++++++++++----
 segv_handler/gnome-segvhanlder.c |    5 +---
 3 files changed, 70 insertions(+), 17 deletions(-)

diff --git a/configure.in b/configure.in
index 46e1401..0d05707 100644
--- a/configure.in
+++ b/configure.in
@@ -73,13 +73,51 @@ AC_CHECK_LIB(elf, elf_version)
 AC_CHECK_HEADERS([libelf.h],[],[AC_MSG_ERROR([libelf.h not found.])])
 AC_CHECK_HEADERS([gelf.h],[],[AC_MSG_ERROR([gelf.h not found.])])
 
-GDK_REQUIRED="x11 gdk-x11-2.0 >= 2.9"
-PKG_CHECK_MODULES(GNOME_SEGV_HANDLER,
-[
-	$GDK_REQUIRED
-])
-AC_SUBST(GNOME_SEGV_HANDLER_CFLAGS)
-AC_SUBST(GNOME_SEGV_HANDLER_LIBS)
+AC_ARG_ENABLE([gtk2-segv-handler],
+        [AC_HELP_STRING([--enable-gtk2-segv-handler],
+                        [Enable crash handler module for Gtk+ 2 [default=auto]])],
+                        enable_gtk2_segv_handler="$enableval", enable_gtk2_segv_handler=auto)
+if test "x$enable_gtk2_segv_handler" != "xno" ; then
+    GDK_REQUIRED="x11 gdk-x11-2.0 >= 2.9"
+    PKG_CHECK_MODULES(GNOME2_SEGV_HANDLER, [$GDK_REQUIRED],
+        [
+          enable_gtk2_segv_handler=yes
+          AC_SUBST(GNOME2_SEGV_HANDLER_CFLAGS)
+          AC_SUBST(GNOME2_SEGV_HANDLER_LIBS)
+        ],
+        [
+          if test "x$enable_gtk2_segv_handler" = "xyes" ; then
+            AC_MSG_ERROR([Gtk+ 2 crash handler module requirements ($GDK_REQUIRED) not met:
+
+$GNOME2_SEGV_HANDLER_PKG_ERRORS])
+          fi
+          enable_gtk2_segv_handler=no
+        ])
+fi
+AM_CONDITIONAL(GNOME2_SEGV_HANDLER, test "x$enable_gtk2_segv_handler" = "xyes")
+
+AC_ARG_ENABLE([gtk3-segv-handler],
+        [AC_HELP_STRING([--enable-gtk3-segv-handler],
+                        [Enable crash handler module for Gtk+ 3 [default=auto]])],
+                        enable_gtk3_segv_handler="$enableval", enable_gtk3_segv_handler=auto)
+if test "x$enable_gtk3_segv_handler" != "xno" ; then
+    GDK_REQUIRED="x11 gdk-x11-3.0"
+    PKG_CHECK_MODULES(GNOME3_SEGV_HANDLER, [$GDK_REQUIRED],
+        [
+          enable_gtk3_segv_handler=yes
+          AC_SUBST(GNOME3_SEGV_HANDLER_CFLAGS)
+          AC_SUBST(GNOME3_SEGV_HANDLER_LIBS)
+        ],
+        [
+          if test "x$enable_gtk3_segv_handler" = "xyes" ; then
+            AC_MSG_ERROR([Gtk+ 3 crash handler module requirements ($GDK_REQUIRED) not met:
+
+$GNOME3_SEGV_HANDLER_PKG_ERRORS])
+          fi
+          enable_gtk3_segv_handler=no
+        ])
+fi
+AM_CONDITIONAL(GNOME3_SEGV_HANDLER, test "x$enable_gtk3_segv_handler" = "xyes")
 
 
 PKG_CHECK_MODULES(GNOME_CRASH, $GTK_REQUIRED)
@@ -127,4 +165,6 @@ Configuration:
 
 	LIBS:	${BUG_BUDDY_LIBS}
 
+	Gtk+ 2 crash handler module:	${enable_gtk2_segv_handler}
+	Gtk+ 3 crash handler module:	${enable_gtk3_segv_handler}
 "
diff --git a/segv_handler/Makefile.am b/segv_handler/Makefile.am
index 5e6c8cb..adf145b 100644
--- a/segv_handler/Makefile.am
+++ b/segv_handler/Makefile.am
@@ -1,9 +1,25 @@
-INCLUDES = -I. -DNDEBUG -D_REENTRANT $(GNOME_SEGV_HANDLER_CFLAGS)
+if GNOME2_SEGV_HANDLER
 
-module_LTLIBRARIES = libgnomesegvhandler.la
+module2_LTLIBRARIES = gtk2/libgnomesegvhandler.la
 
-moduledir = $(libdir)/gtk-2.0/modules
+module2dir = $(libdir)/gtk-2.0/modules
 
-libgnomesegvhandler_la_SOURCES = gnome-segvhanlder.c
-libgnomesegvhandler_la_LIBADD = $(GNOME_SEGV_HANDLER_LIBS)
-libgnomesegvhandler_la_LDFLAGS = -rpath $(moduledir) -module -avoid-version @LDFLAGS@
+gtk2_libgnomesegvhandler_la_SOURCES = gnome-segvhanlder.c
+gtk2_libgnomesegvhandler_la_CFLAGS = -I. -DNDEBUG -D_REENTRANT $(GNOME2_SEGV_HANDLER_CFLAGS)
+gtk2_libgnomesegvhandler_la_LIBADD = $(GNOME2_SEGV_HANDLER_LIBS)
+gtk2_libgnomesegvhandler_la_LDFLAGS = -rpath $(module2dir) -module -avoid-version @LDFLAGS@
+
+endif
+
+if GNOME3_SEGV_HANDLER
+
+module3_LTLIBRARIES = gtk3/libgnomesegvhandler.la
+
+module3dir = $(libdir)/gtk-3.0/modules
+
+gtk3_libgnomesegvhandler_la_SOURCES = gnome-segvhanlder.c
+gtk3_libgnomesegvhandler_la_CFLAGS = -I. -DNDEBUG -D_REENTRANT $(GNOME3_SEGV_HANDLER_CFLAGS)
+gtk3_libgnomesegvhandler_la_LIBADD = $(GNOME3_SEGV_HANDLER_LIBS)
+gtk3_libgnomesegvhandler_la_LDFLAGS = -rpath $(module3dir) -module -avoid-version @LDFLAGS@
+
+endif
diff --git a/segv_handler/gnome-segvhanlder.c b/segv_handler/gnome-segvhanlder.c
index ae3196e..df3ee3d 100644
--- a/segv_handler/gnome-segvhanlder.c
+++ b/segv_handler/gnome-segvhanlder.c
@@ -202,10 +202,7 @@ release_grabs (void)
         gdk_pointer_ungrab(GDK_CURRENT_TIME);
         gdk_keyboard_ungrab(GDK_CURRENT_TIME);
 
-        /* TODO: can we replace this with gdk_x11_ungrab_server ()?
-         * see http://bugzilla.gnome.org/show_bug.cgi?id=550135.
-         */
-	XUngrabServer (GDK_DISPLAY ());
+	XUngrabServer (gdk_x11_get_default_xdisplay ());
 
         gdk_flush();
 
-- 
1.7.5.rc3

